package com.app.security;

import java.io.IOException;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import io.jsonwebtoken.lang.Collections;

@Component
public class JwtFilter extends OncePerRequestFilter {

	@Override
	protected void doFilterInternal(
	        HttpServletRequest request,
	        HttpServletResponse response,
	        FilterChain filterChain)
	        throws ServletException, IOException {

	    String header = request.getHeader("Authorization");

	    if (header != null && header.startsWith("Bearer ")) {

	        String token = header.substring(7);

	        if (JwtUtil.validateToken(token)) {

	        	String email = JwtUtil.extractEmail(token);

	        	UsernamePasswordAuthenticationToken auth =
	        	        new UsernamePasswordAuthenticationToken(email, null, Collections.emptyList());

	        	SecurityContextHolder.getContext().setAuthentication(auth);

	        	System.out.println("Valid Token");
	        	System.out.println("Authenticated User: " + email);

	        } else {

	            System.out.println("Invalid Token");
	        }
	    }

	    filterChain.doFilter(request, response);
	}
}
        
    
